rusoto/services/s3/src/generated.rs

```diff
@@ -19365,22 +19594,28 @@ pub enum HeadObjectError {
     Credentials(CredentialsError),
     /// A validation error occurred.  Details from AWS are provided.
     Validation(String),
+    /// An error occurred parsing the response payload.
+    ParseError(String),
     /// An unknown error occurred.  The raw HTTP response is provided.
-    Unknown(String),
+    Unknown(BufferedHttpResponse),
 }
 
 impl HeadObjectError {
-    pub fn from_body(body: &str) -> HeadObjectError {
-        let reader = EventReader::new(body.as_bytes());
-        let mut stack = XmlResponse::new(reader.into_iter().peekable());
-        find_start_element(&mut stack);
-        match Self::deserialize(&mut stack) {
-            Ok(parsed_error) => match &parsed_error.code[..] {
-                "NoSuchKey" => HeadObjectError::NoSuchKey(String::from(parsed_error.message)),
-                _ => HeadObjectError::Unknown(String::from(body)),
-            },
-            Err(_) => HeadObjectError::Unknown(body.to_string()),
+    pub fn from_response(res: BufferedHttpResponse) -> HeadObjectError {
+        {
+            let reader = EventReader::new(res.body.as_slice());
+            let mut stack = XmlResponse::new(reader.into_iter().peekable());
+            find_start_element(&mut stack);
+            if let Ok(parsed_error) = Self::deserialize(&mut stack) {
+                match &parsed_error.code[..] {
+                    "NoSuchKey" => {
+                        return HeadObjectError::NoSuchKey(String::from(parsed_error.message))
```

Comment:
Based on my experience with the S3 layer, I've never seen this result, only the HeadObjectError::Unknown.

This will allow us to work around the 404 case, the root cause I think for much of the S3 API issues, though I almost wish we could figure out how to resolve that, as the current API is misleading. But this will be much better before we can get to that point.